# Project MBTI

Create a model to predict MBTI personality types from posts

Data format:
    =>type (string): MBTI types
    =>posts (string): text posts 

## Data exploration

In [5]:
import numpy as np
import csv
import pandas as pd

In [17]:
csv_file = csv.reader(open('C:\Users\Gwen\MBTI\mbti_1.csv', 'rb'))

In [37]:
names = ["type", "posts"]

In [72]:
data = np.genfromtxt("C:\Users\Gwen\MBTI\mbti_1.csv", dtype=None, delimiter=',"', names=names, skip_header=1)

In [73]:
print (data)

[ ('INFJ', '\'http://www.youtube.com/watch?v=qsXHcwe3krw|||http://41.media.tumblr.com/tumblr_lfouy03PMA1qa1rooo1_500.jpg|||enfp and intj moments  https://www.youtube.com/watch?v=iz7lE1g4XM4  sportscenter not top ten plays  https://www.youtube.com/watch?v=uCdfze1etec  pranks|||What has been the most life-changing experience in your life?|||http://www.youtube.com/watch?v=vXZeYwwRDw8   http://www.youtube.com/watch?v=u8ejam5DP3E  On repeat for most of today.|||May the PerC Experience immerse you.|||The last thing my INFJ friend posted on his facebook before committing suicide the next day. Rest in peace~   http://vimeo.com/22842206|||Hello ENFJ7. Sorry to hear of your distress. It\'s only natural for a relationship to not be perfection all the time in every moment of existence. Try to figure the hard times as times of growth, as...|||84389  84390  http://wallpaperpassion.com/upload/23700/friendship-boy-and-girl-wallpaper.jpg  http://assets.dornob.com/wp-content/uploads/2010/04/round-home-d

In [105]:
data.dtype.names

('type', 'posts')

Load with panda read_csv to avoid delimiter problem between header and data

In [6]:
data2 = pd.read_csv("C:\Users\Gwen\MBTI\mbti_1.csv", header =0)

In [7]:
print (data2)

      type                                              posts
0     INFJ  'http://www.youtube.com/watch?v=qsXHcwe3krw|||...
1     ENTP  'I'm finding the lack of me in these posts ver...
2     INTP  'Good one  _____   https://www.youtube.com/wat...
3     INTJ  'Dear INTP,   I enjoyed our conversation the o...
4     ENTJ  'You're fired.|||That's another silly misconce...
5     INTJ  '18/37 @.@|||Science  is not perfect. No scien...
6     INFJ  'No, I can't draw on my own nails (haha). Thos...
7     INTJ  'I tend to build up a collection of things on ...
8     INFJ  I'm not sure, that's a good question. The dist...
9     INTP  'https://www.youtube.com/watch?v=w8-egj0y8Qs||...
10    INFJ  'One time my parents were fighting over my dad...
11    ENFJ  'https://www.youtube.com/watch?v=PLAaiKvHvZs||...
12    INFJ  'Joe santagato - ENTP|||ENFJ or  ENTP?   I'm n...
13    INTJ  'Fair enough, if that's how you want to look a...
14    INTP  'Basically this...  https://youtu.be/1pH5c1Jkh...
15    IN

Test columns

In [8]:
data2.columns

Index([u'type', u'posts'], dtype='object')

In [9]:
data2[:0]

Empty DataFrame
Columns: [type, posts]
Index: []

Overview of the dataset

In [10]:
print(data2.shape)

(8675, 2)


We have 8675 raws and 2 columns, which correspond to the dataset in input.

### Frequence of each MBTI type in the dataset

In [11]:
data2.groupby('type').count()

posts
type       
ENFJ    190
ENFP    675
ENTJ    231
ENTP    685
ESFJ     42
ESFP     48
ESTJ     39
ESTP     89
INFJ   1470
INFP   1832
INTJ   1091
INTP   1304
ISFJ    166
ISFP    271
ISTJ    205
ISTP    337

### Create category MBTI list

In [12]:
list(set(data2.type))

['ENFJ',
 'ESFP',
 'INFJ',
 'ESTJ',
 'ISTJ',
 'ENTJ',
 'ISFP',
 'INTJ',
 'ISTP',
 'ENTP',
 'ISFJ',
 'INTP',
 'ESFJ',
 'ESTP',
 'ENFP',
 'INFP']

In [14]:
category = list(data2['type'].unique())

In [15]:
print (category)

['INFJ', 'ENTP', 'INTP', 'INTJ', 'ENTJ', 'ENFJ', 'INFP', 'ENFP', 'ISFP', 'ISTP', 'ISFJ', 'ISTJ', 'ESTP', 'ESFP', 'ESTJ', 'ESFJ']


## Data cleaning

plan:
1) Remove URL
2) Vectorization

### Remove URL using regex

In [17]:
import re

We are applying the regex function on the colum post

In [118]:
for idx, x in enumerate(data2['posts']):
    data2['posts'][idx] = re.sub(r'http([a-z]|[A-Z]|\.|/|\?|=|:|[0-9]|_|-)*','',x)
    #data2['posts'][idx] = re.sub(r'http('(\'|\")','',x)
print(data2['posts'])

0       '||||||enfp and intj moments    sportscenter n...
1       'I'm finding the lack of me in these posts ver...
2       'Good one  _____   |||Of course, to which I sa...
3       'Dear INTP,   I enjoyed our conversation the o...
4       'You're fired.|||That's another silly misconce...
5       '18/37 @.@|||Science  is not perfect. No scien...
6       'No, I can't draw on my own nails (haha). Thos...
7       'I tend to build up a collection of things on ...
8       I'm not sure, that's a good question. The dist...
9       '|||I'm in this position where I have to actua...
10      'One time my parents were fighting over my dad...
11      '|||51 :o|||I went through a break up some mon...
12      'Joe santagato - ENTP|||ENFJ or  ENTP?   I'm n...
13      'Fair enough, if that's how you want to look a...
14      'Basically this...  |||Can I has Cheezburgr?||...
15      'Your comment screams INTJ, bro. Especially th...
16      'some of these both excite and calm me:  BUTTS...
17      'I thi

In [120]:
print(data2)

      type                                              posts  \
0     INFJ  '||||||enfp and intj moments    sportscenter n...   
1     ENTP  'I'm finding the lack of me in these posts ver...   
2     INTP  'Good one  _____   |||Of course, to which I sa...   
3     INTJ  'Dear INTP,   I enjoyed our conversation the o...   
4     ENTJ  'You're fired.|||That's another silly misconce...   
5     INTJ  '18/37 @.@|||Science  is not perfect. No scien...   
6     INFJ  'No, I can't draw on my own nails (haha). Thos...   
7     INTJ  'I tend to build up a collection of things on ...   
8     INFJ  I'm not sure, that's a good question. The dist...   
9     INTP  '|||I'm in this position where I have to actua...   
10    INFJ  'One time my parents were fighting over my dad...   
11    ENFJ  '|||51 :o|||I went through a break up some mon...   
12    INFJ  'Joe santagato - ENTP|||ENFJ or  ENTP?   I'm n...   
13    INTJ  'Fair enough, if that's how you want to look a...   
14    INTP  'Basically th

## Data prepartion: Bag of words 

The objective is to extract numerical features vectors from a text, by tokenizing string (and give to each token an numerical id, using white-space as delimiter), by counting occurence of each token in each posts, by normalizing token (concerning importance of token in occurence/post). And, as most features will be zero (when tokenizing phase), we need to manage stockage space (a possibility is using an inverse document with TF-IDF).

### Vectorization using CountVectorizer of scikit-learn

CountVectorizer is doing in a single class the (1) tokenization and the (2) count occurences

In [126]:
data_for_vec = data2
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
data_counts = count_vect.fit_transform(data_for_vec['posts'])

In [127]:
data_counts.shape

(8675, 110188)

It's give us a 8675 x 110188 sparce matrix, of 955 880 900 element (zero include as we can see above)

In [129]:
data_counts.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 1, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ..., 
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [128]:
print(data_counts)

  (0, 68708)	1
  (0, 92603)	1
  (0, 13880)	1
  (0, 30678)	1
  (0, 66610)	1
  (0, 105427)	1
  (0, 26167)	1
  (0, 47683)	1
  (0, 63943)	1
  (0, 67381)	2
  (0, 72581)	1
  (0, 39250)	1
  (0, 17917)	1
  (0, 10930)	1
  (0, 77066)	1
  (0, 17730)	1
  (0, 58888)	1
  (0, 91080)	1
  (0, 11066)	1
  (0, 108657)	1
  (0, 40249)	1
  (0, 91456)	1
  (0, 79159)	1
  (0, 39237)	1
  (0, 10982)	1
  :	:
  (8674, 71316)	6
  (8674, 54418)	30
  (8674, 98444)	42
  (8674, 73981)	1
  (8674, 29576)	1
  (8674, 68791)	1
  (8674, 39098)	1
  (8674, 67305)	22
  (8674, 97564)	4
  (8674, 58542)	1
  (8674, 109036)	15
  (8674, 63127)	3
  (8674, 70765)	11
  (8674, 41655)	8
  (8674, 71233)	6
  (8674, 109090)	8
  (8674, 51675)	17
  (8674, 59557)	2
  (8674, 97122)	39
  (8674, 17153)	2
  (8674, 47412)	2
  (8674, 105974)	5
  (8674, 98783)	1
  (8674, 69749)	12
  (8674, 12542)	34


In [131]:
count_vect.vocabulary_.get('moments')

65914

Here an example, we have 65914 occurences of the word "moments" in the posts. 

Now we have (1) tokenize the posts content and (2) count the words occurencies, but it's better to work with frequencies.

Currently we are using only dictionnary of 1-grams (individual words) because default value for the function CountVectorizer() is ngram_range=(1, 1). But we can take 2-grams (using parameter ngram_range=(1, 2) in the function) if we want to enrich the dictionnary (we have also group of two words), but it also taking much more place. So I will choose for the moment to use only 1-gram (and maybe enrich later if necessary).

### TF-IDF

To avoid space stockage issues, I will use the TF-IDF (term frequency-inverse document frequency) methods, which will allow for example (1:tf) to devide each occurences of a word in a post by total number of words in this post (that give new features names tf). This method also allow (2:tf-idf) to give less weight (importance) on words which appear in many posts of the corpus and automatically that give less information than words that appear in a less sample of posts.

$ tfidf(t,d) = tf(t,d) * idf(t) $

In [136]:
from sklearn.feature_extraction.text import TfidfTransformer
tfidf_transformer = TfidfTransformer()
data_tfidf = tfidf_transformer.fit_transform(data_counts)
data_tfidf.shape

(8675, 110188)

With the function "fit_tranform()" we are doing in one time the fit() which fit the estimator to data (1:tf) and the transform() which transform the matrix of occurences (data_counts) in tf-idf matrix with frequencies (2:tf-idf).

### Train and Test set

I will now devide the dataset in trainset (75%) and testset (25%) to execute models. I will use a specific train/test split function given by scikit-learn.

In [138]:
from sklearn.model_selection import train_test_split
trainset, testset = train_test_split(data2, test_size=0.25)

In [139]:
print(trainset)

      type                                              posts  \
7246  INFJ  'It depends on the group but in the past, with...   
3526  INFJ  'Dear classmates (just a few),  It's not cool ...   
3078  INFJ  'With maybe a belt like this  ...|||Thanks! I ...   
482   INTJ  'I used to love Nutella, so I decided to buy a...   
1564  ISTJ  'CallSignOWL's post reminded me of a question....   
6199  INFJ  'To never become like my mom and having to hav...   
1086  INFP  'Ok, I'll go first. I'm a 29 year old INFP mal...   
4624  INFJ  'Not always. It depends on the severity of the...   
3858  ENFP  A lot of this is assuming that we're talking a...   
3389  INFJ  'Welcome! My favorite type is INTJs, you guys ...   
3587  INFP  'I'm curious too|||Omg this right here lol. I ...   
2600  INFP  'All of you guys saying Just cut them out of y...   
8042  INFP  Is it ever possible to have the perfect life? ...   
5351  INFJ  'infjs do this too. I have a system for gettin...   
4711  INFP  'oh boy- star

In [140]:
trainset.count()

type     6506
posts    6506
post     6506
test     6506
dtype: int64

A trainset with 75% of the dataframe (8675 raws)

In [141]:
testset.count()

type     2169
posts    2169
post     2169
test     2169
dtype: int64

A testset with 25% of the dataframe (8675 raws)